tests/operators/test_grid_sampling_op.py

```diff
@@ -0,0 +1,268 @@
+
+
+def test_grid_sampling_op_gradcheck_grid_forward():
+    rng = RandomGenerator(0).float64_tensor
+    grid = rng((2, 1, 2, 2), -0.8, 0.8).requires_grad_(True)
+    u = rng((1, 1, 3, 5))
+    gradcheck(lambda grid, u: GridSamplingOp(grid, input_shape=SpatialDimension(1, 3, 5))(u), (grid, u), fast_mode=True)
+
+
+def test_grid_sampling_op_gradcheck_x_adjoint():
+    rng = RandomGenerator(0).float64_tensor
+    grid = rng((2, 1, 2, 2), -0.8, 0.8).requires_grad_(True)
+    v = rng((2, 1, 1, 2))
+    gradcheck(
+        lambda grid, v: GridSamplingOp(grid, input_shape=SpatialDimension(1, 2, 3)).adjoint(v),
+        (grid, v),
+        fast_mode=True,
+    )
+
+
+def test_grid_sampling_op_gradcheck_grid_adjoint():
+    rng = RandomGenerator(0).float64_tensor
+    grid = rng((2, 1, 2, 2), -0.8, 0.8).requires_grad_(True)
+    v = rng((2, 1, 1, 2))
+    gradcheck(
+        lambda grid, v: GridSamplingOp(grid, input_shape=SpatialDimension(1, 2, 3)).adjoint(v),
+        (grid, v),
+        fast_mode=True,
+    )
+
+
+def test_grid_sampling_op_errormsg_gridlastdim():
+    "Test if error message on wrong last dim is raised."
+    grid = torch.ones(1, 2, 3, 4)
+    with pytest.raises(ValueError, match='last dimension'):
+        _ = GridSamplingOp(grid, SpatialDimension(1, 1, 1))
+
+
+def test_grid_sampling_op_errormsg_gridndims():
+    "Test if error message on missing batch dim is raised."
+    grid = torch.ones(1, 1, 1, 3)  # 3d
+    with pytest.raises(ValueError, match='batch z y x 3'):
+        _ = GridSamplingOp(grid, SpatialDimension(1, 1, 1))
+
+    grid = torch.ones(1, 1, 2)  # 2d
+    with pytest.raises(ValueError, match='batch y x 2'):
+        _ = GridSamplingOp(grid, SpatialDimension(1, 1, 1))
+
+
+def test_grid_sampling_op_errormsg_cubic3d():
+    "Test if error for 3D cubic is raised."
+    grid = torch.ones(1, 1, 1, 1, 3)  # 3d
+    with pytest.raises(NotImplementedError, match='cubic'):
+        _ = GridSamplingOp(grid, SpatialDimension(1, 1, 1), interpolation_mode='bicubic')
+
+
+def test_grid_sampling_op_errormsg_complexgrid():
+    "Test if error for complex grid is raised."
+    grid = torch.ones(1, 1, 1, 1, 3) + 0j
+    with pytest.raises(ValueError, match='real'):
+        _ = GridSamplingOp(grid, SpatialDimension(1, 1, 1))
+
+
+def test_grid_sampling_op_errormsg_gridrange():
+    "Test if warnding for grid values outside [-1,1] is raised"
+    grid = torch.ones(1, 1, 1, 1, 3)
+
+    grid.ravel()[0] = 1.01
+    with pytest.warns(UserWarning, match='range'):
+        _ = GridSamplingOp(grid, SpatialDimension(1, 1, 1))
+
+    grid.ravel()[0] = -1.01
+    with pytest.warns(UserWarning, match='range'):
+        _ = GridSamplingOp(grid, SpatialDimension(1, 1, 1))
+
+    grid.ravel()[0] = -1.0
+    _ = GridSamplingOp(grid, SpatialDimension(1, 1, 1))
+
+    grid.ravel()[0] = 1.0
+    _ = GridSamplingOp(grid, SpatialDimension(1, 1, 1))
+
+
+def test_grid_sampling_op_errormsg_inputdim():
```

Comment:
can this be rewritten like `test_grid_sampling_op_batchdims`?

Reply:
I prefer not to. I had it parameterize, but due to warning/raise/works and really simple tests it was more complicated to read...

Reply:
I could do it as many seperate tests if prefered..

Reply:
I understand that parameterization might lead to more verbose or complex code but the problem I see here is that the test can fail for a number of reasons. Depending on when it fails the majority of tests might not actually be run which is not good practice for unit testing.

Reply:
split into different tests